### PR TITLE
Native messaging

### PR DESCRIPTION
**Pull request by openbots-aw**
- Added ChromeNativeMessaging Commands
- Added ChromeNativeMessaging Server Project
- Added Chrome Extension
- Added ChromeNativeMessaging Recorder in Selenium Element Action

### CODE REVIEW

File: OpenBots.Commands.NativeMessagingAutomation/Class1.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿using System;
+
+namespace OpenBots.Commands.NativeMessagingAutomation
+{
+    public class Class1
```
Comment (openbots-ff) [inline OpenBots.Commands.NativeMessagingAutomation/Class1.cs:5]: is this class necessary?

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+		[SampleUsage("MyChromeBrowserInstance")]
+		[Remarks("Failure to enter the correct instance name or failure to first call the **Create Browser** command will cause an error.")]
+		[CompatibleTypes(new Type[] { typeof(Process) })]
+		public string v_InstanceName { get; set; }
+
+		[Required]
+		[DisplayName("Element Search Parameter")]
+		[Description("Use the Element Recorder to generate a listing of potential search parameters." +
+					"Select the specific search type(s) that you want to use to isolate the element on the web page.")]
+		[SampleUsage("XPath : \"//*[@id=\"features\"]/div[2]/div/h2/div[\" + var1 + \"]/div\"" +
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs:42]:
```suggestion
		[SampleUsage("XPath : \"//*[@id='features']/div[2]/div/h2/div[\" + var1 + \"]/div\"" +
```

Comment (openbots-ff) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs:42]: this change should be made in all commands

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+			string searchParameterValue = (from rw in v_NativeSearchParameters.AsEnumerable()
+										   where rw.Field<string>("Enabled") == "True"
+										   select rw.Field<string>("Parameter Value")).FirstOrDefault();
+
+			return base.GetDisplayValue() + $" [Clear Text by {searchParameterName}" +
+											$" '{searchParameterValue}' - Instance Name '{v_InstanceName}']";
+		}
+		private void ActionParametersGridViewHelper_MouseEnter(object sender, EventArgs e)
+		{
+			//SeleniumAction_SelectionChangeCommitted(null, null);
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs:164]: please remove this event handler if you don't intend to use it.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+{
+	[Serializable]
+	[Category("Native Messaging Commands")]
+	[Description("This command closes a native web browser session.")]
+	public class NativeMessagingCloseBrowserCommand : ScriptCommand
+	{
+		[Required]
+		[DisplayName("Browser Instance Name")]
+		[Description("Enter the unique instance that was specified in the **Create Browser** command.")]
+		[SampleUsage("\"MyChromeBrowserInstance\"")]
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs:25]: Instance names are not strings so they should not have quotation marks around them

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		[Remarks("Failure to enter the correct instance name or failure to first call the **Create Browser** command will cause an error.")]
+		[CompatibleTypes(new Type[] { typeof(Process) })]
+		public string v_InstanceName { get; set; }
+
+		public NativeMessagingCloseBrowserCommand()
+		{
+			CommandName = "NativeMessagingCloseBrowserCommand";
+			SelectionName = "Close Browser";
+			v_InstanceName = "\"DefaultChromeBrowser\"";
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs:34]: the sample shouldn't have quotation marks either

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+		[DisplayName("Browser Instance Name")]
+		[Description("Enter the unique instance that was specified in the **Create Browser** command.")]
+		[SampleUsage("\"MyChromeBrowserInstance\"")]
+		[Remarks("Failure to enter the correct instance name or failure to first call the **Create Browser** command will cause an error.")]
+		[CompatibleTypes(new Type[] { typeof(Process) })]
+		public string v_InstanceName { get; set; }
+
+		[DisplayName("URL (Optional)")]
+        [Description("Enter the URL that you want the selenium instance to navigate to.")]
+        [SampleUsage("https://mycompany.com/orders || {vURL}")]
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs:40]: please add quotation marks to the URL and remove the {} from {vURL}. These changes should be made to all samples.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+			if(chromeProcessCount > 0)
+            {
+				while (process.HasExited == false)
+				{
+					System.Threading.Thread.Sleep(100);
+					process.Refresh();
+				}
+			}
+			//Delay 7 seconds
+			System.Threading.Thread.Sleep(7000);
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs:74]: Please shorten this

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingElementExistsCommand.cs.cs [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+		{
+			string searchParameterName = (from rw in v_NativeSearchParameters.AsEnumerable()
+										  where rw.Field<string>("Enabled") == "True"
+										  select rw.Field<string>("Parameter Name")).FirstOrDefault();
+
+			string searchParameterValue = (from rw in v_NativeSearchParameters.AsEnumerable()
+										   where rw.Field<string>("Enabled") == "True"
+										   select rw.Field<string>("Parameter Value")).FirstOrDefault();
+
+			return base.GetDisplayValue() + $" [Element Exists for {searchParameterName}" +
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingElementExistsCommand.cs.cs:172]: This command has an output so you should include that in the display value the same way it's done in other commands with instances and outputs

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingGetAttributeCommand.cs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+						 "\n\tLink Text: \"https://www.mylink.com/\"")]
+		[Remarks("If multiple parameters are enabled, an attempt will be made to find the element(s) that match(es) all the selected parameters.")]
+		[Editor("ShowElementHelper", typeof(UIAdditionalHelperType))]
+		[CompatibleTypes(new Type[] { typeof(string) })]
+		public DataTable v_NativeSearchParameters { get; set; }
+
+		[Required]
+		[DisplayName("Attribute Name")]
+		[Description("Enter the attribute name to be set.")]
+		[SampleUsage("\"Hello World\" || vAttribute")]
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingGetAttributeCommand.cs:58]: could you provide a more relevant sample here?

File: OpenBots.Studio/Resources/Extension/OB Extension/ContentListener.js [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+		var responseObj;
+		var elementValue = "";
+		console.log(sender.tab ?
+			"from a content script:" + sender.tab.url :
+			"from the extension");
+		if (request.responseType == "getelement") {
+			document.addEventListener("click", onClickEvent);
+			document.addEventListener("mouseover", mouseOverEvent);
+			document.addEventListener("mouseout", mouseOutEvent);
+			//sendResponse({targetHTML: clickPoints});
```
Comment (openbots-ff) [inline OpenBots.Studio/Resources/Extension/OB Extension/ContentListener.js:12]: should this be removed?

File: OpenBots.NativeServer/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("OpenBots.NativeServer")]
```
Comment (openbots-ff) [inline OpenBots.NativeServer/Properties/AssemblyInfo.cs:8]: please modify all the metadata in this file to match the other projects

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+		[Remarks("If multiple parameters are enabled, an attempt will be made to find the element(s) that match(es) all the selected parameters.")]
+		[Editor("ShowElementHelper", typeof(UIAdditionalHelperType))]
+		[CompatibleTypes(new Type[] { typeof(string) })]
+		public DataTable v_NativeSearchParameters { get; set; }
+
+		[JsonIgnore]
+		[Browsable(false)]
+		private DataGridView _searchParametersGridViewHelper;
+
+		public NativeMessagingLeftClickCommand()
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs:59]: Could you please combine the click commands?

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+
+namespace OpenBots.Commands.NativeMessaging
+{
+	[Serializable]
+	[Category("Native Messaging Commands")]
+	[Description("This command performs left click on specified element in chrome.")]
+	public class NativeMessagingLeftClickCommand : ScriptCommand
+	{
+		[Required]
+		[DisplayName("Browser Instance Name")]
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs:31]:
```suggestion
		[DisplayName("Chrome Browser Instance Name")]
```

Comment (openbots-ff) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs:31]: please make this change in all commands to distinguish between these instances and the selenium ones

**Review by openbots-ff — CHANGES_REQUESTED** [review]

File: OpenBots.Commands.NativeMessagingAutomation/Class1.cs (hunk 1 above)
Comment (openbots-aw) [reply OpenBots.Commands.NativeMessagingAutomation/Class1.cs:5]: Its removed.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs (hunk 3 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingClearTextCommand.cs:164]: Done.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs (hunk 4 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs:25]: Done.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs (hunk 5 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCloseBrowserCommand.cs:34]: Done.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs (hunk 6 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs:40]: Done.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs (hunk 7 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs:74]: Changed to 3 seconds.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingElementExistsCommand.cs.cs (hunk 8 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingElementExistsCommand.cs.cs:172]: Done.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingGetAttributeCommand.cs (hunk 9 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingGetAttributeCommand.cs:58]: Done.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs (hunk 12 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs:59]: Currently only left click is working, and I have tried to make the other clicks work but JavaScript is not supporting. I will have to find the alternate ways to support other clicks. Until then we cannot combine them.

File: OpenBots.NativeServer/Properties/AssemblyInfo.cs (hunk 11 above)
Comment (openbots-aw) [reply OpenBots.NativeServer/Properties/AssemblyInfo.cs:7]: Updated.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs (hunk 13 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingLeftClickCommand.cs]: Updated.

File: OpenBots.Studio/Resources/Extension/OB Extension/ContentListener.js (hunk 10 above)
Comment (openbots-aw) [reply OpenBots.Studio/Resources/Extension/OB Extension/ContentListener.js]: I have removed the comment.

File: OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs (hunk 7 above)
Comment (openbots-ff) [reply OpenBots.Commands/OpenBots.Commands.UIAutomation/OpenBots.Commands.NativeMessaging/NativeMessagingCreateBrowserCommand.cs]: I didn't mean the time lol. I meant the namespace before it

**Review by openbots-ff — APPROVED** [review]

**Review by openbots-ch — DISMISSED** [review]

**Review by openbots-ch — APPROVED** [review]

**Pull request merged**